proxyd/consensus_tracker.go

```diff
@@ -21,101 +26,308 @@ type ConsensusTracker interface {
 
-	ct.latestBlockNumber = blockNumber
+	ct.state.Latest = blockNumber
 }
 
 func (ct *InMemoryConsensusTracker) GetSafeBlockNumber() hexutil.Uint64 {
 	defer ct.mutex.Unlock()
 	ct.mutex.Lock()
 
-	return ct.safeBlockNumber
+	return ct.state.Safe
 }
 
 func (ct *InMemoryConsensusTracker) SetSafeBlockNumber(blockNumber hexutil.Uint64) {
 	defer ct.mutex.Unlock()
 	ct.mutex.Lock()
 
-	ct.safeBlockNumber = blockNumber
+	ct.state.Safe = blockNumber
 }
 
 func (ct *InMemoryConsensusTracker) GetFinalizedBlockNumber() hexutil.Uint64 {
 	defer ct.mutex.Unlock()
 	ct.mutex.Lock()
 
-	return ct.finalizedBlockNumber
+	return ct.state.Finalized
 }
 
 func (ct *InMemoryConsensusTracker) SetFinalizedBlockNumber(blockNumber hexutil.Uint64) {
 	defer ct.mutex.Unlock()
 	ct.mutex.Lock()
 
-	ct.finalizedBlockNumber = blockNumber
+	ct.state.Finalized = blockNumber
 }
 
-// RedisConsensusTracker uses a Redis `client` to store and retrieve consensus, async-safe
+// RedisConsensusTracker store and retrieve in a shared Redis cluster, with leader election
 type RedisConsensusTracker struct {
 	ctx          context.Context
 	client       *redis.Client
-	backendGroup string
+	namespace    string
+	backendGroup *BackendGroup
+
+	redlock           *redsync.Mutex
+	lockPeriod        time.Duration
+	heartbeatInterval time.Duration
+
+	leader     bool
+	leaderName string
+
+	// holds the state collected by local pollers
+	local *InMemoryConsensusTracker
+
+	// holds a copy of the remote shared state
+	// when leader, updates the remote with the local state
+	remote *InMemoryConsensusTracker
+}
+
+type RedisConsensusTrackerOpt func(cp *RedisConsensusTracker)
+
+func WithLockPeriod(lockPeriod time.Duration) RedisConsensusTrackerOpt {
+	return func(ct *RedisConsensusTracker) {
+		ct.lockPeriod = lockPeriod
+	}
+}
+
+func WithHeartbeatInterval(heartbeatInterval time.Duration) RedisConsensusTrackerOpt {
+	return func(ct *RedisConsensusTracker) {
+		ct.heartbeatInterval = heartbeatInterval
+	}
 }
+func NewRedisConsensusTracker(ctx context.Context,
+	redisClient *redis.Client,
+	bg *BackendGroup,
+	namespace string,
+	opts ...RedisConsensusTrackerOpt) ConsensusTracker {
 
-func NewRedisConsensusTracker(ctx context.Context, r *redis.Client, namespace string) ConsensusTracker {
-	return &RedisConsensusTracker{
+	tracker := &RedisConsensusTracker{
 		ctx:          ctx,
-		client:       r,
-		backendGroup: namespace,
+		client:       redisClient,
+		backendGroup: bg,
+		namespace:    namespace,
+
+		lockPeriod:        30 * time.Second,
+		heartbeatInterval: 2 * time.Second,
+		local:             NewInMemoryConsensusTracker().(*InMemoryConsensusTracker),
+		remote:            NewInMemoryConsensusTracker().(*InMemoryConsensusTracker),
+	}
+
+	for _, opt := range opts {
+		opt(tracker)
+	}
+
+	return tracker
+}
+
+func (ct *RedisConsensusTracker) Init() {
+	go func() {
+		for {
+			timer := time.NewTimer(ct.heartbeatInterval)
+			ct.stateHeartbeat()
+
+			select {
+			case <-timer.C:
```

Comment:
```suggestion
			case <-timer.C:
			    continue
```

Added `continue` to make this clearer